File: api/models/pre_assessment.py
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field


SourceType = Literal["pdf_md", "ppt_md", "note_md", "other_md"]
ReadinessLevel = Literal["ready", "needs_preparation", "not_now"]
ConfidenceLevel = Literal["high", "medium", "low"]
EvidenceSource = Literal["material", "profile", "both"]
ProfileSuggestionType = Literal[
    "possible_gap", "learning_preference", "workflow_signal", "strength"
]


class PreAssessmentRequest(BaseModel):
    source_type: SourceType
    title: str = Field(..., min_length=1, max_length=300)
    full_content: str = Field(..., min_length=1)
    selected_excerpt: Optional[str] = Field(default=None)
    excerpt_context_before: Optional[str] = Field(default=None)
    excerpt_context_after: Optional[str] = Field(default=None)
    user_goal_hint: Optional[str] = Field(default=None, max_length=1000)
    profile_overrides: Dict[str, Any] = Field(default_factory=dict)


class ReadinessAssessment(BaseModel):
    level: ReadinessLevel
    confidence: ConfidenceLevel
    summary: str = Field(..., min_length=1, max_length=400)


class BlockingGap(BaseModel):
    concept: str = Field(..., min_length=1, max_length=120)
    why_it_blocks: str = Field(..., min_length=1, max_length=400)
    evidence_source: EvidenceSource
    suggested_preparation: str = Field(..., min_length=1, max_length=300)


class CoreFrame(BaseModel):
    what_this_material_is_really_about: str = Field(
        ..., min_length=1, max_length=400
    )
    key_axes: List[str] = Field(default_factory=list, max_length=5)
    what_is_foundational: List[str] = Field(default_factory=list, max_length=4)
    what_is_detail_or_later: List[str] = Field(default_factory=list, max_length=4)


class ReadingStrategy(BaseModel):
    focus_first: List[str] = Field(default_factory=list, max_length=3)
    skim_or_skip_for_now: List[str] = Field(default_factory=list, max_length=3)
    questions_to_keep_in_mind: List[str] = Field(default_factory=list, max_length=3)


class EvidenceNote(BaseModel):
    claim: str = Field(..., min_length=1, max_length=200)
    basis: str = Field(..., min_length=1, max_length=300)


class ProfileUpdateSuggestion(BaseModel):
    type: ProfileSuggestionType
    content: str = Field(..., min_length=1, max_length=300)


class PreAssessmentResponse(BaseModel):
    readiness: ReadinessAssessment
    blocking_gaps: List[BlockingGap] = Field(default_factory=list, max_length=3)
    core_frame: CoreFrame
    reading_strategy: ReadingStrategy
    evidence_notes: List[EvidenceNote] = Field(default_factory=list, max_length=3)
    profile_update_suggestions: List[ProfileUpdateSuggestion] = Field(
        default_factory=list, max_length=3
    )

File: src/types/pre-assessment.ts
export type SourceType = "pdf_md" | "ppt_md" | "note_md" | "other_md";
export type ReadinessLevel = "ready" | "needs_preparation" | "not_now";
export type ConfidenceLevel = "high" | "medium" | "low";
export type EvidenceSource = "material" | "profile" | "both";
export type ProfileSuggestionType =
  | "possible_gap"
  | "learning_preference"
  | "workflow_signal"
  | "strength";

export interface PreAssessmentRequest {
  source_type: SourceType;
  title: string;
  full_content: string;
  selected_excerpt?: string;
  excerpt_context_before?: string;
  excerpt_context_after?: string;
  user_goal_hint?: string;
  profile_overrides: Record<string, unknown>;
  provider?: string;
  model?: string;
  api_key?: string;
  base_url?: string;
}

export interface ReadinessAssessment {
  level: ReadinessLevel;
  confidence: ConfidenceLevel;
  summary: string;
}

export interface BlockingGap {
  concept: string;
  why_it_blocks: string;
  evidence_source: EvidenceSource;
  suggested_preparation: string;
}

export interface CoreFrame {
  what_this_material_is_really_about: string;
  key_axes: string[];
  what_is_foundational: string[];
  what_is_detail_or_later: string[];
}

export interface ReadingStrategy {
  focus_first: string[];
  skim_or_skip_for_now: string[];
  questions_to_keep_in_mind: string[];
}

export interface EvidenceNote {
  claim: string;
  basis: string;
}

export interface ProfileUpdateSuggestion {
  type: ProfileSuggestionType;
  content: string;
}

export interface PreAssessmentResponse {
  readiness: ReadinessAssessment;
  blocking_gaps: BlockingGap[];
  core_frame: CoreFrame;
  reading_strategy: ReadingStrategy;
  evidence_notes: EvidenceNote[];
  profile_update_suggestions: ProfileUpdateSuggestion[];
}
